core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -286,11 +286,14 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
     tryOrElse(that.map(Right(_)), ZIO.succeedLeft)
 
   private final def tryOrElse[R1 <: R, E2, B](that: => ZIO[R1, E2, B], succ: A => ZIO[R1, E2, B]): ZIO[R1, E2, B] =
-    (self.tag: @switch) match {
-      case ZIO.Tags.Fail => that
-
-      case _ => new ZIO.Fold[R1, E, E2, A, B](self, _ => that, succ)
-    }
+    new ZIO.Fold[R1, E, E2, A, B](
+      self,
+      _.defects.headOption match {
+        case Some(d) => ZIO.die(d)
```

Comment:
This is lossy right now in the sense that it discards the rest of the defects and failures. I'd like to improve on this but we can't return a `Cause[E]` if there are failures because of the type signature.

The best thing we can do is discard failures and only return defects. However, this would still discard information, as any `Both` or `Then` terms containing a failure must be changed to a `Die`.

Reply:
Also, we can't demand that `E2 >: E` as that would not allow patterns like `mayFail <> alwaysSucceeds`.

Reply:
This is messy and will involve lots of allocations. If it's only used for `orElse`, maybe that's OK. 

Did you look at `Cause#failureOrCause`? I think it will let you keep all the information

Reply:
It's insufficient unfortunately, because a failure doesn't guarantee the absence of a defect.

I think I can just pattern match directly on cause to look for a defect. That way we won't allocate (but we'll still have to traverse a tree of causes).

Reply:
done